docs/architecture/smart-contracts.md

```diff
@@ -0,0 +1,154 @@
+# Smart contracts
+
+The Koinos Blockchain Framework is a bare bones minimal blockchain implementation that is fully customization through the use of smart contracts. Smart contracts can implement both feature rich
+decentralized applications and core system functionality. The information in this documented is intended to be SDK agnostic and explain the core functionality of the Koinos Virtual Machine and its
+implementation.
+
+Because Koinos uses [Fizzy](https://github.com/wasmx/fizzy) for its WASM (WebAssembly) virtual machine, the blockchain is agnostic to the language the smart has been written -- this allows
+support for a multitude of programming languages for development.
+
+Smart contracts come in two flavors: user and system. User contracts have a basic set of feature that allow the developer to
+write decentralized applications. System contracts have all the features available to user contracts, with the additional ability of access the system contract space.
+
+## Contract space
+
+Each smart contract has access to an `object_space`, which essentially defines a key value store. The `object_space` is defined as follows:
+
+```proto
+message object_space {
+   bool system = 1;
+   bytes zone = 2;
+   uint32 id = 3;
+}
+```
+
+### User space
+In the case of a user contract `system` will be set to `false`, the `zone` will be set to the bytes of the public address of the contract, and `id` is used as a unique identifier distinguishing between multiple key value stores. A contract may
+have different key value stores by incrementing the `id`.
+
+### System space
+The system contract is special in that it may access a _global_ space denoted by `system` being `true` and the `zone` being set to an empty byte array, otherwise known as `kernel` space. This allows for system contracts to read and manipulate a pool
+of shared key value stores.
+
+## Entry points
+
+To inform the blockchain which function you are calling within a smart contract an entry point is provided. The Koinos Blockchain Framework will take the request, whether it is a read or write, load up the contract and pass the entry point as a parameter.
+The main function will then instantiate the contract class or execute procedurally and execute the corresponding code whether it a member function or inline code. Entry points come in two flavors: read and write.
+
+### Read-only entry points
+
+Read-only entry points are often used to implement "getter" functions and can be called outside of a transaction. Most commonly, the read only entry points are called using the RPC method `read_contract`. Smart contract code that is executed
+in read-only mode is prevented from writing to any key value store and will be abruptly trapped and a permission denied exception will propagate if any attempt is made.
```

Comment:
`will be abruptly trapped, and a permission denied exception will propagate if any attempt is made.`

Add that comma to make this sentence much easier to read